pennylane_lightning/core/src/utils/UtilLinearAlg.hpp

```diff
@@ -0,0 +1,91 @@
+
+/**
+ * @file
+ * LAPACK wrapper functions declaritions.
+ */
+#pragma once
+
+#include <algorithm>
+#include <complex>
+#include <vector>
+
+namespace {
+extern "C" {
+// LAPACK routine for complex Hermitian eigensystems
+extern void cheev_(const char *jobz, const char *uplo, const int *n,
+                   std::complex<float> *a, const int *lda, float *w,
+                   std::complex<float> *work, const int *lwork, float *rwork,
+                   int *info);
+extern void zheev_(const char *jobz, const char *uplo, const int *n,
+                   std::complex<double> *a, const int *lda, double *w,
+                   std::complex<double> *work, const int *lwork, double *rwork,
+                   int *info);
+}
+} // namespace
+
+namespace Pennylane::Util {
+
+/**
+ * @brief Decompose Hermitian matrix into diagonal matrix and unitaries
+ *
+ * @tparam T Data type.
+ *
+ * @param n Number of columns.
+ * @param lda Number of rows.
+ * @param Ah Hermitian matrix to be decomposed.
+ * @param eigenVals eigenvalue results.
+ * @param unitaries unitary result.
+ */
+template <typename T>
+void compute_diagonalizing_gates(int n, int lda,
+                                 const std::vector<std::complex<T>> &Ah,
+                                 std::vector<T> &eigenVals,
+                                 std::vector<std::complex<T>> &unitary) {
+    eigenVals.clear();
+    eigenVals.resize(n);
+    unitary.clear();
+    unitary = std::vector<std::complex<T>>(n * n, {0, 0});
```

Comment:
do you need to clear the unitary matrix?
```suggestion
    unitary = std::vector<std::complex<T>>(n * n, {0, 0});
```

Reply:
Done!